packages/core/saved-objects/core-saved-objects-migration-server-internal/src/document_migrator/pipelines/upgrade_pipeline.ts

```diff
@@ -7,38 +7,59 @@
  */
 
 import Boom from '@hapi/boom';
+import { cloneDeep } from 'lodash';
 import Semver from 'semver';
 import type { SavedObjectUnsanitizedDoc } from '@kbn/core-saved-objects-server';
-import { ActiveMigrations, Transform, TransformType } from './types';
-import { maxVersion } from './utils';
+import { ActiveMigrations, Transform, TransformType } from '../types';
+import type { MigrationPipeline, MigrationPipelineResult } from './types';
+import {
+  coreVersionTransformTypes,
+  applyVersion,
+  assertValidCoreVersion,
+  maxVersion,
+} from './utils';
 
 function isGreater(a?: string, b?: string) {
   return !!a && (!b || Semver.gt(a, b));
 }
 
-export class DocumentMigratorPipeline {
-  additionalDocs = [] as SavedObjectUnsanitizedDoc[];
-
-  constructor(
-    public document: SavedObjectUnsanitizedDoc,
-    private migrations: ActiveMigrations,
-    private kibanaVersion: string,
-    private convertNamespaceTypes: boolean
-  ) {}
+export class DocumentUpgradePipeline implements MigrationPipeline {
+  private additionalDocs = [] as SavedObjectUnsanitizedDoc[];
+  private document: SavedObjectUnsanitizedDoc;
+  private originalDoc: SavedObjectUnsanitizedDoc;
+  private migrations: ActiveMigrations;
+  private kibanaVersion: string;
+  private convertNamespaceTypes: boolean;
+  private targetTypeVersion: string;
+
+  constructor({
+    document,
+    migrations,
+    kibanaVersion,
+    convertNamespaceTypes,
+    targetTypeVersion,
+  }: {
+    document: SavedObjectUnsanitizedDoc;
+    migrations: ActiveMigrations;
+    kibanaVersion: string;
+    convertNamespaceTypes: boolean;
+    targetTypeVersion?: string;
+  }) {
+    this.originalDoc = document;
+    this.document = cloneDeep(document);
+    this.migrations = migrations;
+    this.kibanaVersion = kibanaVersion;
+    this.convertNamespaceTypes = convertNamespaceTypes;
+    this.targetTypeVersion = targetTypeVersion || migrations[document.type]?.latestVersion.migrate;
+  }
 
   protected *getPipeline(): Generator<Transform> {
     while (this.hasPendingTransforms()) {
-      const { type } = this.document;
-
       for (const transform of this.getPendingTransforms()) {
         yield transform;
 
-        if (type !== this.document.type) {
-          // In the initial implementation, all the transforms for the new type should be applied.
-          // And at the same time, documents with `undefined` in `typeMigrationVersion` are treated as the most recent ones.
-          // This is a workaround to get into the loop again and apply all the migrations for the new type.
-          this.document.typeMigrationVersion = '';
-          break;
+        if (this.document.type !== this.originalDoc.type) {
+          throw new Error(`Changing a document's type during a migration is not supported.`);
```

Comment:
This is (finally) no longer possible, as it would be a real problem for ZDT.

Note that no one ever used this functionality anyway.